python/ray/data/_internal/logical/interfaces.py

```diff
@@ -0,0 +1,89 @@
+from typing import List
+
+
+class Operator:
+    """Abstract class for operators.
+
+    Operators are stateful and non-serializable; they live on the driver side of the
+    Dataset only.
+    """
+
+    def __init__(self, name: str, input_dependencies: List["Operator"]):
+        self._name = name
+        self._input_dependencies = input_dependencies
+        for x in input_dependencies:
+            assert isinstance(x, Operator), x
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    @property
+    def input_dependencies(self) -> List["Operator"]:
+        """List of operators that provide inputs for this operator."""
+        assert hasattr(
+            self, "_input_dependencies"
+        ), "Operator.__init__() was not called."
+        return self._input_dependencies
+
+    def __repr__(self) -> str:
+        if self.input_dependencies:
+            out_str = ", ".join([str(x) for x in self.input_dependencies])
+            out_str += " -> "
+        else:
+            out_str = ""
+        out_str += f"{self.__class__.__name__}[{self._name}]"
+        return out_str
+
+    def __str__(self) -> str:
+        return repr(self)
+
+
+class LogicalOperator(Operator):
+    """Abstract class for logical operators.
+
+    A logical operator describes transformation, and later is converted into
+    physical operator.
+    """
+
+    def __init__(self, name: str, input_dependencies: List["LogicalOperator"]):
+        super().__init__(name, input_dependencies)
+        for x in input_dependencies:
+            assert isinstance(x, LogicalOperator), x
+
+    def set_input_dependencies(self, input_dependencies: List["LogicalOperator"]):
+        self._input_dependencies = input_dependencies
+
+    def get_physical_dag(self):
+        """Get the corresponding DAG of physical operators.
+
+        Returns:
+            DAG of physical operators as List[PhysicalOperator].
+        """
+        raise NotImplementedError
+
+
+class Rule:
+    """Abstract class for optimization rule."""
+
+    def apply(dag: Operator) -> Operator:
```

Comment:
Shall we need actual DAG (i.e. QueryPlan) class? That seems more natural (than current DAG is-a Operator). And when we have more rules, there will be common graph editing operations needed, which can be shared via this QueryPlan class.

Reply:
I was doing it in my initial commit - https://github.com/ray-project/ray/pull/31731/commits/bdec4cffc233c9edb36f35cc62d0d4426187d2e2 , but got feedback here - https://github.com/ray-project/ray/pull/31731#discussion_r1073927564 . I am thinking we can introduce the Plan class later, and I agree with you that common graph traversal operation is definitely needed.